test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+// will still have the correct output values from the af_* function, even though
+// the array was initially reordered.
+void genReorderedArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    // This reorder combination will not move data around, but will simply
+    // call modDims and modStrides (see src/api/c/reorder.cpp).
+    // The output will be checked if it is still correct even with the
+    // modified dims and strides "hack" with no data movement
+    unsigned reorder_0 = 0;
+    unsigned reorder_1 = 2;
+    unsigned reorder_2 = 1;
+
+    // Shape the output array such that the reordered output array will have
+    // the correct dimensions that the test asks for (i.e. must match dims arg)
+    af::dim4 out_dims(dims[reorder_0], dims[reorder_1], dims[reorder_2]);
+    metadata.initArrays(out_dims, ty);
+    metadata.setOutputArray(af::reorder(metadata.out_arr,
+                                        reorder_0, reorder_1, reorder_2));
+}
+
+// Partner function of testWriteToOutputArray. This generates the "special"
+// array that testWriteToOutputArray will use to check if the af_* function
+// correctly uses an existing array as its output
+void genTestOutputArray(af_array *out_ptr,
+                        const unsigned ndims, const dim_t *dims,
+                        const af::dtype ty,
+                        TestOutputArrayInfo* metadata) {
+    af::dim4 arr_dims(ndims, dims);
+    switch (metadata->getOutputArrayType()) {
+    case NULL_ARRAY:
+        genNullArray(*metadata, arr_dims, ty);
+        break;
+    case FULL_ARRAY:
+        genRegularArray(*metadata, arr_dims, ty);
+        break;
+    case SUB_ARRAY:
+        genSubArray(*metadata, arr_dims, ty);
+        break;
+    case REORDERED_ARRAY:
+        genReorderedArray(*metadata, arr_dims, ty);
+        break;
+    }
+    *out_ptr = metadata->getOutputArrayC();
+}
+
+// Partner function of genTestOutputArray. This uses the same "special"
+// array that genTestOutputArray generates, and checks whether the
+// af_* function wrote to that array correctly
+::testing::AssertionResult
+testWriteToOutputArray(std::string gold_name, std::string result_name,
+                       const af_array gold, const af_array out,
+                       TestOutputArrayInfo *metadata) {
+    // In the case of NULL_ARRAY, the output array starts out as null.
+    // After the af_* function is called, it shouldn't be null anymore
+    if (metadata->getOutputArrayType() == NULL_ARRAY) {
+        if (out == 0) {
+            return ::testing::AssertionFailure()
+                << "Output af_array " << result_name << " is null";
+        }
+        metadata->setOutputArrayC(out);
+    }
+    // For every other case, must check if the af_array generated by
+    // genTestOutputArray was used by the af_* function as its output array
+    else {
+        if (metadata->getOutputArrayC() != out) {
```

Comment:
`metadata->getOutputArray().get()`

Reply:
`metadata->getOutputArray().get()` will not equal `out`. The latter is the output of `af_retain_array` in `genTestOutputArray`, and that is going to be different from the result of `af::array::get()`. This is one of the reasons why I currently store the output of the retain call as a member of `TestOutputArrayInfo`.